crates/hotshot/types/src/stake_table.rs

```diff
@@ -40,4 +47,123 @@ impl<K: SignatureKey> StakeTableEntry<K> {
     }
 }
 
-// TODO(Chengyu): add stake table snapshot here
+#[derive(Debug, Clone, PartialEq, Eq, Hash, Default)]
+pub struct HSStakeTable<TYPES: NodeType>(pub Vec<PeerConfig<TYPES>>);
+
+impl<TYPES: NodeType> From<Vec<PeerConfig<TYPES>>> for HSStakeTable<TYPES> {
+    fn from(peers: Vec<PeerConfig<TYPES>>) -> Self {
+        Self(peers)
+    }
+}
+
+impl<TYPES: NodeType> std::ops::Deref for HSStakeTable<TYPES> {
+    type Target = Vec<PeerConfig<TYPES>>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl<TYPES: NodeType> std::ops::DerefMut for HSStakeTable<TYPES> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+#[inline]
+/// A helper function to compute the quorum threshold given a total amount of stake.
+pub fn one_honest_threshold(total_stake: U256) -> U256 {
+    total_stake / U256::from(3) + U256::from(1)
+}
+
+#[inline]
+fn u256_to_field(amount: U256) -> CircuitField {
+    let amount_bytes: [u8; 32] = amount.to_le_bytes();
+    CircuitField::from_le_bytes_mod_order(&amount_bytes)
+}
+
+impl<TYPES: NodeType> std::iter::IntoIterator for HSStakeTable<TYPES> {
+    type Item = PeerConfig<TYPES>;
+    type IntoIter = std::vec::IntoIter<PeerConfig<TYPES>>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
```

Comment:
i think once you have the `Deref` implemented, it's automatic that any methods inner struct (which is vec) has, should be available in the wrapper struct. so this is unnecessary.

Reply:
It's still necessary because `.into_iter()` takes ownership. `Deref` cannot take care of it.